wsmux/session.go

```diff
@@ -0,0 +1,334 @@
+package wsmux
+
+import (
+	"bytes"
+	"encoding/binary"
+	"io/ioutil"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/gorilla/websocket"
+)
+
+/*
+TODO: Add ping and pong handlers
+*/
+
+const (
+	defaultQueueSize            = 20
+	defaultStreamQueueSize      = 20
+	defaultKeepAliveInterval    = 30 * time.Second
+	defaultStreamAcceptDeadline = 30 * time.Second
+)
+
+// Session implements net.Listener. Allows creating and acception multiplexed streams over ws
+type Session struct {
+	mu       sync.Mutex
+	streams  map[uint32]*stream
+	streamCh chan *stream
+	writes   chan frame
```

Comment:
Why note make a function `s.write(frame) error` that locks and sends?

That way we know if the connection had an error while sending... And we know it has been sent before we move forward..

With a channel we just know that the sending thread is about to send it... But not if an error occured... Etc..